lib/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,318 @@
+		 *
+		 * @since 6.7.0
+		 * @var WP_Block_Templates_Registry|null
+		 */
+		private static $instance = null;
+
+		/**
+		 * Registers a block template.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string|WP_Block_Template $template_name Block template name including namespace, or alternatively
+		 *                                                a complete WP_Block_Template instance. In case a WP_Block_Template
+		 *                                                is provided, the $args parameter will be ignored.
+		 * @param string                   $template_type Template type, either `wp_template` or `wp_template_part`.
+		 * @param array                    $args          Optional. Array of block template arguments.
+		 * @return WP_Block_Template|false The registered block template on success, or false on failure.
+		 */
+		public function register( $template_name, $template_type, $args = array() ) {
+
+			$template = null;
+			if ( $template_name instanceof WP_Block_Template ) {
+				$template      = $template_name;
+				$template_name = $template->name;
+			}
+
+			if ( ! is_string( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must be strings.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( 'wp_template' !== $template_type && 'wp_template_part' !== $template_type ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block templates need to be of `wp_template` or `wp_template_part` type.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( preg_match( '/[A-Z]+/', $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must not contain uppercase characters.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			$name_matcher = '/^[a-z0-9-]+\/\/[a-z0-9-]+$/';
+			if ( ! preg_match( $name_matcher, $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( $this->is_registered( $template_type, $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					/* translators: %s: Template name. */
+					sprintf( __( 'Template "%s" is already registered.', 'gutenberg' ), $template_name ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( ! $template ) {
+				$theme_name = get_stylesheet();
+				$slug       = isset( $args['slug'] ) ? $args['slug'] : explode( '//', $template_name )[1];
+
+				$template                 = new WP_Block_Template();
+				$template->id             = $theme_name . '//' . $slug;
+				$template->theme          = $theme_name;
+				$template->plugin         = isset( $args['plugin'] ) ? $args['plugin'] : '';
+				$template->author         = null;
+				$template->content        = isset( $args['content'] ) ? $args['content'] : '';
+				$template->source         = 'plugin';
```

Comment:
Right, I found this in some places so far, like `is-template-revertable.js` and `is-template-removable.js`, where I needed to update the checks to account for plugin templates.

Your proposal it's not 100% clear to me. Would you keep `source` and `origin` as `theme` for plugin-registered templates and use the `plugin` property to identify the templates which come from plugins?

Reply:
I remember working on this in the past and might be able to shed some light. I remember I found that the templates API had very little support for plugins and had been made originally only for theme provided templates (that's why there are fields like `theme` rather than something more generic), but at the same time there were already plugins providing templates via one of the available filters.

I remember `source` was being set to a value of 'plugin' by plugins to match how themes used it, but wasn't suitable for anything because as soon as a template is customized it's value changes to `custom`. So I added `origin` as a way to show the original source value even after a template has been customized. I don't think it needs to be set to anything when registering templates, as core I think core copies the `source` to the `origin`. You might need to confirm that.

I think themes were also adding theme names/ids/slugs to the `theme` field, as there was nothing else available. The design team wanted to show the theme title in one of the columns, so where that requirement was from.

I believe I also added `author` support for custom templates added by users, as that was also missing. That's all I recall, it was a while ago!

I found the PR here - [Add author support to templates and template parts](https://github.com/WordPress/wordpress-develop/pull/1937#top)

Reply:
It's still not clear to me why we need the extra prop and can't rely on `origin===plugin`.

Reply:
Makes sense, yes. So my new questions would be..
Is `plugin` value in the register function a specific value for a plugin? Does it mean it should be identical to this first part `plugin_name//template_slug`? 

>Sidenote: we could get rid of the plugin property from the API and infer the plugin from the template name (plugin_name//template_slug), but I think that's not what we are discussing here, right?

So would that also mean that if we keep the `plugin` prop, we could get rid of the prefix of `(plugin_name//template_slug)`? If both of them are not really needed, do we need them both? Is there any other reason for keeping this info in two places?

Reply:
Thanks for all the good questions, @ntsekouras! Yes, the API had some duplication between the template name and the args themselves, we could infer the plugin name and the template slug.

In e5f735c7a9637efa4aa782c8f27ba8b6bba32944 I refactored the registry so it accepts template names without a prefix and requires the `plugin` property to be defined. Internally, the registry still stores templates in the format `plugin_name//template_name` to avoid conflicts between plugins. The reason is that if we want to support plugins registering templates used in the WP hierarchy, we need to handle the case of two plugins registering templates with the same slug without erroing.

I also took the chance to get rid of the `slug` property from the API. Now we assume the slug will match the `template_name`. I updated the testing steps in the PR description accordingly.

Reply:
> I refactored the registry so it accepts template names without a prefix and requires the `plugin` property to be defined.

Thinking about it twice, I'm leaning towards adding back the prefix and removing the `plugin` property. This way we can [keep the signatures of the `gutenberg_register_template()` and `gutenberg_unregister_template()`](https://github.com/WordPress/gutenberg/pull/61577/files#diff-2ef3fa362b7d011f82a7b0fa1178406290f8cfe0dbf92a561f596ebaff92c6eaR26-R38) more similar, what do you think?

Reply:
I'm marking this conversation as resolved based on the comments here: https://github.com/WordPress/gutenberg/pull/61577#issuecomment-2252451616. _tl;dr_: we will add back the namespace and remove the `plugin` property.